cpp/src/arrow/util/thread_pool.h

```diff
@@ -301,8 +301,10 @@ class ARROW_EXPORT SerialExecutor : public Executor {
   /// approach is to use a stop token to cause the generator to exhaust early.
   template <typename T>
   static Iterator<T> IterateGenerator(
-      internal::FnOnce<Result<std::function<Future<T>()>>(Executor*)> initial_task) {
+      internal::FnOnce<Result<std::function<Future<T>()>>(Executor*)> initial_task,
+      internal::SerialExecutor** serial_executor_out = NULLPTR) {
```

Comment:
It is a bit strange. At least so far I didn't find a better way to expose, given that the [`SerialExecutor` constructor is private](https://github.com/apache/arrow/blob/daacbcc4c5f0e435b2158896584a4385dbf38986/cpp/src/arrow/util/thread_pool.h#L370-L371). I needed to expose it to [send the stop-token to it](https://github.com/apache/arrow/pull/35953/files#diff-70e843181eb977ed11ad131f79a0f54a0bf9e30a0cb752d440658ee0cc16cc34R1076).

Reply:
The `StopProducing` approach ([noted here](https://github.com/apache/arrow/pull/35953/#issuecomment-1589892510)) worked in an internal experiment. Side note: the pre-PR version also had a call to `StopProducing`, in the destructor of `BatchConverter`, but it was too late - a hang happens earlier.

Reply:
If I read the code correctly - it seems like BatchConverter destructor is called when the PlanReader gets destructed (since it is a member of the batch_iterator which is a member of the PlanReader) (related code: https://github.com/apache/arrow/blob/e53db939bfad2f20e332172ab4f453add1dc680d/cpp/src/arrow/acero/exec_plan.cc#L1028)

This means that the destructor doesn't gets called during `Close()` / `iterator_.reset()`, which would explain why it didn't work before. @westonpace Does that sound correct?

Reply:
That sounds right though an additional point is that `Close` hanged in the loop for draining the remaining batches, so the destructor was never reached.